### PR TITLE
plan: optimize the filter condition push down logic when do index double read

### PR DESCRIPTION
**Pull request by lamxTyler**
When we do index double read, some filter condition can be pushed down to the index request.

### CODE REVIEW

**Comment by shenli** [comment]
@lamxTyler Please add some unit test cases.

File: plan/refiner.go [hunk 1]
```diff
@@ -128,6 +128,58 @@ func removeAccessConditions(conditions, accessConds []expression.Expression) []e
 	return conditions
 }
 
+func judgeIndexExpression(arg expression.Expression, indexColumns []*model.IndexColumn, pKName model.CIStr) bool {
```
Comment (shenli) [inline plan/refiner.go]: Add comment for this function.

File: plan/refiner.go [hunk 2]
```diff
@@ -128,6 +128,58 @@ func removeAccessConditions(conditions, accessConds []expression.Expression) []e
...
+		for _, col := range indexColumns {
+			if col.Name.L == c.ColName.L && col.Length == types.UnspecifiedLength {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func judgeIndexCondition(condition expression.Expression, indexColumns []*model.IndexColumn, pKName model.CIStr) bool {
```
Comment (coocood) [inline plan/refiner.go]: Use `check` instead of `judge`.
And we should just use one function, do recursive call, because an argument may be a scalar function.

Comment (hanfei1991) [inline plan/refiner.go]: For this function, we should extract all columns from the expression and check for every column. There is a function called expression.extractColumns.

File: plan/physical_plan_builder.go [hunk 3]
```diff
@@ -245,7 +245,12 @@ func (p *DataSource) convert2IndexScan(prop *requiredProperty, index *model.Inde
 		if client != nil {
 			memDB := infoschema.IsMemoryDB(p.DBName.L)
 			if !memDB && client.SupportRequestType(kv.ReqTypeIndex, 0) {
-				is.ConditionPBExpr, is.conditions, newSel.Conditions = expressionsToPB(sc, newSel.Conditions, client)
+				var indexConditions, pbConditions []expression.Expression
+				indexConditions, newSel.Conditions = detachIndexFilterConditions(newSel.Conditions, is.Index.Columns, is.Table)
+				is.IndexPBExpr, is.conditions, indexConditions = expressionsToPB(sc, indexConditions, client)
+				is.ConditionPBExpr, pbConditions, newSel.Conditions = expressionsToPB(sc, newSel.Conditions, client)
+				newSel.Conditions = append(newSel.Conditions, indexConditions...)
+				is.conditions = append(is.conditions, pbConditions...)
```
Comment (hanfei1991) [inline plan/physical_plan_builder.go]: You can split is.conditions into two parts: IndexFilterConditions and TableFilterConditions. And show them in explain result.

**Comment by lamxTyler** [comment]
PTAL @shenli @coocood @hanfei1991

File: executor/executor_test.go [hunk 4]
```diff
@@ -827,6 +827,17 @@ func (s *testSuite) TestIndexScan(c *C) {
 	tk.MustExec("INSERT INTO tab1 VALUES(1,37,20.85,30,10.69)")
 	result = tk.MustQuery("SELECT pk FROM tab1 WHERE ((col3 <= 6 OR col3 < 29 AND (col0 < 41)) OR col3 > 42) AND col1 >= 96.1 AND col3 = 30 AND col3 > 17 AND (col0 BETWEEN 36 AND 42)")
 	result.Check(testkit.Rows())
+	tk.MustExec("drop table if exists tab1")
```
Comment (hanfei1991) [inline executor/executor_test.go]: This case can't actually cover you change. Because the "a" may used to calcuate range and result in access conditions.
you should create index (a,b) and involved a column c which is not in any index.

File: executor/explain_test.go [hunk 5]
```diff
@@ -53,7 +53,8 @@ func (s *testSuite) TestExplain(c *C) {
     "push down info": {
         "limit": 0,
         "access conditions": null,
-        "filter conditions": null
+        "index filter conditions": null,
+        "table filter conditions": null
```
Comment (hanfei1991) [inline executor/explain_test.go]: add a test that both index filter and table filter are not null.

File: executor/executor_test.go (hunk 4 above)
Comment (lamxTyler) [inline executor/executor_test.go]: But if "a" is used in access conditions, then "pk" will be in index filter conditions

**Comment by hanfei1991** [comment]
LGTM

File: plan/physical_plan_builder.go [hunk 6]
```diff
@@ -245,7 +245,11 @@ func (p *DataSource) convert2IndexScan(prop *requiredProperty, index *model.Inde
 		if client != nil {
 			memDB := infoschema.IsMemoryDB(p.DBName.L)
 			if !memDB && client.SupportRequestType(kv.ReqTypeIndex, 0) {
-				is.ConditionPBExpr, is.conditions, newSel.Conditions = expressionsToPB(sc, newSel.Conditions, client)
+				var indexConditions []expression.Expression
+				indexConditions, newSel.Conditions = detachIndexFilterConditions(newSel.Conditions, is.Index.Columns, is.Table)
```
Comment (coocood) [inline plan/physical_plan_builder.go]: Define a new temporary variable `tableConditons` instead of use `newSel.Conditon` is better.

File: plan/refiner.go [hunk 7]
```diff
@@ -128,6 +128,48 @@ func removeAccessConditions(conditions, accessConds []expression.Expression) []e
...
+	var pKName model.CIStr
+	if table.PKIsHandle {
+		for _, colInfo := range table.Columns {
+			if mysql.HasPriKeyFlag(colInfo.Flag) {
+				pKName = colInfo.Name
+				break
+			}
+		}
+	}
+	var indexConditions, filterConditions []expression.Expression
```
Comment (coocood) [inline plan/refiner.go]: filterConditions -> tableConditions

**Comment by coocood** [comment]
LGTM
